packages/cli/src/modules/insights/__tests__/insights-compaction.service.test.ts

```diff
@@ -317,22 +320,30 @@ if (dbType === 'sqlite' && !globalConfig.database.sqlite.poolSize) {
 
 		describe('compactionSchedule', () => {
 			test('compaction is running on schedule', async () => {
+				// ARRANGE
 				jest.useFakeTimers();
+				const insightsCompactionService = new InsightsCompactionService(
+					mock<InsightsByPeriodRepository>(),
+					mock<InsightsRawRepository>(),
+					mock<InsightsConfig>({
+						compactionIntervalMinutes: 60,
+					}),
+					mockLogger(),
+				);
+				// spy on the compactInsights method to check if it's called
+				const compactInsightsSpy = jest.spyOn(insightsCompactionService, 'compactInsights');
+
 				try {
-					// ARRANGE
-					const insightsCompactionService = Container.get(InsightsCompactionService);
 					insightsCompactionService.startCompactionTimer();
 
-					// spy on the compactInsights method to check if it's called
-					const compactInsightsSpy = jest.spyOn(insightsCompactionService, 'compactInsights');
-
 					// ACT
 					// advance by 1 hour and 1 minute
 					jest.advanceTimersByTime(1000 * 60 * 61);
 
 					// ASSERT
 					expect(compactInsightsSpy).toHaveBeenCalledTimes(1);
 				} finally {
+					insightsCompactionService.stopCompactionTimer();
```

Comment:
I added this to make sure the instance is not kept around by the timer not being cleared.

A thing I notices is that there is no way to cleanly stop the compaction. I was removing the mocks above and ended up with the compaction still running in the background when when the db was disconnected.

I imagine this would also happen when shutting down n8n propper. We'd stop the timer and then shutdown, but if a compaction is currently running it would fail. Since it's running in a transaction that not causing damage, but the customer would see the error in the logs and could think something is wrong and contact support.

Do you think it's worth fixing this as well in another PR? e.g. have `stopCompactionTimer` return a promise the resolves when the current compaction finishes?

Reply:
Yes indeed, good catch. I created a ticket: https://linear.app/n8n/issue/PAY-2807/wait-for-ongoing-compaction-to-finish-on-shutdown
I'll try to work on it on friday